src/zjs_ble.c

```diff
@@ -285,52 +283,45 @@ static jerry_value_t zjs_ble_write_attr_call_function_return(const jerry_value_t
     return ZJS_UNDEFINED;
 }
 
-static void zjs_ble_write_attr_call_function(struct zjs_callback *cb)
+static void zjs_ble_write_c_callback(void *handle)
 {
-    ble_handle_t *mycb;
-    ble_characteristic_t *chrc;
-    mycb = CONTAINER_OF(cb, ble_handle_t, zjs_cb);
-    chrc = CONTAINER_OF(mycb, ble_characteristic_t, write_cb);
+    ble_characteristic_t *chrc = (ble_characteristic_t *)handle;
+    ble_handle_t *cb = &chrc->write_cb;
 
     jerry_value_t rval;
     jerry_value_t args[4];
     jerry_value_t func_obj;
 
-    if (mycb->buffer && mycb->buffer_size > 0) {
-        jerry_value_t buf_obj = zjs_buffer_create(mycb->buffer_size);
+    args[0] = jerry_create_null();
+    if (cb->buffer && cb->buffer_size > 0) {
+        jerry_value_t buf_obj = zjs_buffer_create(cb->buffer_size);
 
         if (buf_obj) {
            zjs_buffer_t *buf = zjs_buffer_find(buf_obj);
 
            if (buf &&
                buf->buffer &&
-               buf->bufsize == mycb->buffer_size) {
-               memcpy(buf->buffer, mycb->buffer, mycb->buffer_size);
-               args[0]= buf_obj;
-           } else {
-               args[0] = jerry_create_null();
+               buf->bufsize == cb->buffer_size) {
+               memcpy(buf->buffer, cb->buffer, cb->buffer_size);
            }
-        } else {
-            args[0] = jerry_create_null();
+
+           jerry_release_value(args[0]);
+           args[0]= buf_obj;
         }
-    } else {
-        args[0] = jerry_create_null();
     }
-
-    args[1] = jerry_create_number(mycb->offset);
+    args[1] = jerry_create_number(cb->offset);
+    // TODO: support withoutResponse flag
     args[2] = jerry_create_boolean(false);
-    func_obj = jerry_create_external_function(zjs_ble_write_attr_call_function_return);
-    jerry_set_object_native_handle(func_obj, (uintptr_t)chrc, NULL);
+    func_obj = jerry_create_external_function(zjs_ble_write_callback_function);
+    jerry_set_object_native_handle(func_obj, (uintptr_t)handle, NULL);
     args[3] = func_obj;
 
-    rval = jerry_call_function(mycb->zjs_cb.js_callback, chrc->chrc_obj, args, 4);
+    rval = jerry_call_function(cb->js_callback, chrc->chrc_obj, args, 4);
     if (jerry_value_has_error_flag(rval)) {
-        PRINT("zjs_ble_write_attr_call_function: failed to call onWriteRequest function\n");
+        PRINT("zjs_ble_write_c_callback: failed to call onWriteRequest function\n");
     }
 
     jerry_release_value(args[0]);
-    jerry_release_value(args[1]);
-    jerry_release_value(args[2]);
```

Comment:
argp[1] and arg[2] are integer and boolean, so they do not need to be released, but I'll put them back in for the sake in the case the jerryscript api changes in the future